File: sdks/java/ToolsetClient.java
package toolset;

import java.io.IOException;
import java.net.URI;
import java.net.http.HttpClient;
import java.net.http.HttpRequest;
import java.net.http.HttpResponse;
import java.time.Duration;

public class ToolsetClient {
  private final String baseUrl;
  private final HttpClient http;
  private String token;
  private String apiKey;

  public ToolsetClient(String baseUrl) {
    this.baseUrl = baseUrl.replaceAll("/+$", "");
    this.http = HttpClient.newBuilder().connectTimeout(Duration.ofSeconds(10)).build();
  }

  public ToolsetClient withToken(String token) {
    this.token = token;
    return this;
  }

  public ToolsetClient withApiKey(String apiKey) {
    this.apiKey = apiKey;
    return this;
  }

  private HttpRequest.Builder baseRequest(String path) {
    HttpRequest.Builder b = HttpRequest.newBuilder().uri(URI.create(this.baseUrl + path)).timeout(Duration.ofSeconds(60));
    b.header("content-type", "application/json");
    if (this.token != null && !this.token.isEmpty()) b.header("authorization", "Bearer " + this.token);
    if (this.apiKey != null && !this.apiKey.isEmpty()) b.header("x-api-key", this.apiKey);
    return b;
  }

  public String listTools() throws IOException, InterruptedException {
    HttpRequest req = baseRequest("/v1/tools").GET().build();
    HttpResponse<String> resp = http.send(req, HttpResponse.BodyHandlers.ofString());
    if (resp.statusCode() >= 400) throw new IOException("http " + resp.statusCode() + ": " + resp.body());
    return resp.body();
  }

  public String invoke(String toolName, String jsonBody) throws IOException, InterruptedException {
    HttpRequest req = baseRequest("/v1/tools/" + toolName + ":invoke").POST(HttpRequest.BodyPublishers.ofString(jsonBody)).build();
    HttpResponse<String> resp = http.send(req, HttpResponse.BodyHandlers.ofString());
    if (resp.statusCode() >= 400) throw new IOException("http " + resp.statusCode() + ": " + resp.body());
    return resp.body();
  }
}

File: sdks/python/toolset_client.py
from __future__ import annotations

import requests


class ToolsetClient:
    def __init__(self, base_url: str, *, token: str | None = None, api_key: str | None = None):
        self.base_url = base_url.rstrip("/")
        self.token = token
        self.api_key = api_key

    def _headers(self) -> dict[str, str]:
        h = {"content-type": "application/json"}
        if self.token:
            h["authorization"] = f"Bearer {self.token}"
        if self.api_key:
            h["x-api-key"] = self.api_key
        return h

    def get_token(self, *, client_id: str, client_secret: str) -> dict:
        r = requests.post(
            f"{self.base_url}/v1/oauth/token",
            json={"client_id": client_id, "client_secret": client_secret},
            timeout=10,
        )
        r.raise_for_status()
        return r.json()

    def list_tools(self) -> dict:
        r = requests.get(f"{self.base_url}/v1/tools", headers=self._headers(), timeout=10)
        r.raise_for_status()
        return r.json()

    def invoke(self, tool_name: str, *, input: dict, context: dict | None = None, options: dict | None = None) -> dict:
        payload = {"input": input, "context": context or {}, "options": options or {}}
        r = requests.post(
            f"{self.base_url}/v1/tools/{tool_name}:invoke",
            headers=self._headers(),
            json=payload,
            timeout=60,
        )
        r.raise_for_status()
        return r.json()

